Refactor robustness methods

aulemahal opened the pull request:
<!--Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [ ] Tests for the changes have been added (for bug fixes / features)
  - [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGES.rst has been updated (with summary of main changes)
  - [ ] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Add `ipcc-advanced` as a test to `ensembles.change_significance`. This corresponds to the simpler alternative of approach C as described in the Cross-Chapter Box 1 of the [IPCC Atlas (AR6, WG1)](https://www.cambridge.org/core/books/climate-change-2021-the-physical-science-basis/atlas/24E1C016DBBE4725BDFBC343695DE7DB).
* Add two examples in the "Ensembles" notebook to show how xclim can be used to get the IPCC-recommended hatching masks.
* Add a very simple `detrend` function.

### Does this PR introduce a breaking change?
No, but it might, see below.

### Other information:
As you can see in the examples, the IPCC decouples the "change significance" from the "sign agreement", while xclim does not. This make the second example more complex than expected.

I thought of :
- Changing the output of `change_significance` so that `pos_frac` is independent of change significance
- Adding a third output so we have: `significant_change_frac, significant_positive_change_frac, positive_change_frac`.
- Adding a keyword argument to switch between the two modes.
First two are breaking changes while the thirds feels overly heavy for something this simple...

huard commented on docs/notebooks/ensembles.ipynb:
```diff
@@ -275,6 +275,188 @@
+    "- $\\gamma = \\sqrt{2}\\cdot1.645\\cdot\\sigma_{20yr}$ where $\\sigma_{20yr}$ is the standard deviation of 20-year means, computed from non-overlapping periods in the pre-industrical control runs, after detrending with a quadratic fit.\n",
+    "- $\\gamma = \\sqrt{\\frac{2}{20}}\\cdot1.645\\cdot\\sigma_{1yr}$ where $\\sigma_{1yr}$ is the interanual standard deviation computed over the linearly detrended reference period.\n",
+    "\n",
+    "The first method involves data that is not always easily available. When such data is at hand, the xclim function could be used with a custom computation of the $\\gamma$ (`gamma`) threshold:\n",
```
```suggestion
    "The first method involves data from pre-industrial control simulations that are not always easily available. When such data is at hand, the xclim function could be used with a custom computation of the $\\gamma$ (`gamma`) threshold:\n",
```

huard replied:
The second sentence is unclear. Does it mean the first approach is not implemented ?

aulemahal replied:
The `gamma` computation is indeed not implemented for the case where pre-industrial control simulations are available. I'll rephrase.

huard commented on docs/notebooks/ensembles.ipynb:
````diff
@@ -275,6 +275,188 @@
+    "chg_frac, pos_frac = ensembles.change_significance(fut, ref, test='threshold', abs_thresh=gamma)\n",
+    "```\n",
+    "\n",
+    "The second method is more accessible and xclim implements it directly. However, `pos_frac` output of xclim is the fraction of the members showing _significant_ positive change, while the IPCC defines the agreement separately from the change significance. So we run the function twice."
````
That's not super clear.

aulemahal replied:
What do you think of the apparent complexity of this ?  See main comment, but the double call comes from the our definition of `pos_frac` that is different from the IPCC's. This could be modified, but would be breaking.

huard replied:
Here I'm not suggestion to change the code itself, just the explanations. In fact, You may not have to give so much information in the notebook. I suggest to focus on the method that's actually implemented (don't mention the first method), and explain how it differs from the IPCC's in a bit more sentences. 

The "So we run the function twice." is a bit cryptic.

huard commented on xclim/ensembles/_robustness.py:
```diff
@@ -49,7 +50,7 @@ def change_significance(  # noqa: C901
         `fut` and `ref` must be of the same type (Dataset or DataArray). If they are
         Dataset, they must have the same variables (name and coords).
-    test : {'ttest', 'welch-ttest', 'mannwhitney-utest', 'brownforsythe-test', 'threshold', None}
+    test : {'ttest', 'welch-ttest', 'mannwhitney-utest', 'brownforsythe-test', 'ipcc-advanced', 'threshold', None}
```
I'd be more specific regarding naming, qqch comme `ipcc-ar6-c`.

huard replied:
Also, let's say we wanted to implement later the first gamma calculation with preindustrial data, would we have to break anything? Should we call it `ipcc-ar6-c-hist` and later implement `ipcc-ar6-c-pi` ? or something ?

aulemahal replied:
Agreed to change `ipcc-advanced` to `ipcc-ar6-c`. 

Yes it would be breaking as this function would now need 3 inputs instead 2. : future, reference (for the deltas) and preindustrial (for `gamma`).

In fact, we could implement both as another function : `gamma(ref, input_type='pi'/'hist')` (or maybe a better name). The `ipcc-ar6-c` test in the main function would simply call `gamma(ref, 'hist')` while a user with PI data would call:
```python3
change_significance(fut, ref, test='threshold', abs_thresh=gamma(pi_data, 'pi'))
```
Does that make sense ?

huard replied:
I _think_ having a third input (preindustrial) would be more user friendly. I'd implement the `gamma` function as a private function.

huard commented on xclim/ensembles/_robustness.py:
```diff
@@ -323,7 +331,12 @@ def mwu_wrapper(f, r):  # This specific test can't manage an all-NaN slice
+        # Detrend ref
+        refy = ref.resample(time="YS").mean()
+        ref_detrended = detrend(refy, dim="time", deg=1)
+        gamma = np.sqrt(2 / 20) * 1.645 * ref_detrended.std("time")
```
In the NB, you say we can provide a custom gamma, but I don't see it here.

aulemahal replied:
See other comment, but indeed, I'll make the distinction clearer in the notebook.

huard commented on xclim/ensembles/_robustness.py:
```diff
@@ -111,13 +112,19 @@ def change_significance(  # noqa: C901
         Change is considered significative if the absolute delta exceeds a given threshold (absolute or relative).
+      'ipcc-advanced' :
+        An approximation of the "advanced approach" used in the IPCC Atlas and described in :cite:t:`ipccatlas_ar6wg1`.
+        Change is considered significant if the delta exceeds :math:`1.645\sqrt{\frac{2}{10}}\sigma_{ref'}$ where $\sigma_{ref'}` is
```
Isn't it 2/20 ?

huard reviewed (commented):
- Looks good. 
- The nb text assumes the reader's familiarity with the topic and the IPCC approach. I suggest to take a more pedagogic approach. 
- I don't see a test ; )

aulemahal commented:
@huard After our discussions and some thinking, I rewrote the function completely. 

I was not satisfied with the output of `change_significance` so I rewrote it as `robustness_fractions`. The former still works but it will warn the user that it is deprecated. Also added `robustness_categories` to help with making the actual map.

Finally, I simplified the notebook a lot. May be too much...

I'll add tests tomorrow.

aulemahal commented:
@RondeauG , would you have time to look at the new functions ? And may be @mccrayc, if you have time.

I'm simply looking for a review of the general idea and usage, as shown in the example : https://xclim--1514.org.readthedocs.build/en/1514/notebooks/ensembles.html#Change-significance-and-model-agreement

RondeauG commented on docs/notebooks/ensembles.ipynb:
```diff
@@ -164,6 +164,7 @@
     "# Set display to HTML style (for fancy output)\n",
```
<p>Quelques détails de syntaxe ou typo:</p><ul><li><code>surimposed</code> --&gt; <code>superimposed</code>   </li><li>change significance : whether most of the ensemble members project <strong>a statistically significant climate change signal, </strong>in comparison to their internal variability.</li><li>There are<s> be</s> 5 different members in the ensemble</li></ul><p> </p><blockquote>(models agree and all show significant change)</blockquote><p>Je dirais plutôt <code>enough show a significant change</code>. L'IPCC utilise un seuil de X% (60?).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/xclim/pull/1514/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

huard reviewed (approved):
Ça me semble beaucoup plus clair comme ça. Pas testé en vrai, juste lu en diagonale.

RondeauG commented:
This looks pretty good! `robustness_categories` will be very useful.
